002-es-modules.md

```diff
@@ -0,0 +1,540 @@
+
+* implement interoperability for ESM and Node's existing module system
+* create a **Registry Object** (see WHATWG section below) compatible with
+  the [WHATWG Loader](http://whatwg.github.io/loader/) Registry and/or [WHATWG module map](https://html.spec.whatwg.org/multipage/webappapis.html#module-map)
+
+## 1. Purpose
+
+1. Allow a common module syntax for Browser and Server.
+2. Allow a common registry for inspection by Browser and Server
+   environments/tools.
+    * These will most likely be represented by metaproperties like
+      `import.context`, but the spec is not yet fully in place.
+
+## 2. Related
+
+[ECMA262](tc39.github.io/ecma262/) discusses the syntax and semantics of
+related syntax, and introduces:
+
+### 2.1. Types
+
+* **[ModuleRecord]
+(https://tc39.github.io/ecma262/#sec-abstract-module-records)**
+    - Defines the list of imports via `[[ImportEntry]]`.
+    - Defines the list of exports via `[[ExportEntry]]`.
+
+* **[ModuleNamespace]
+(https://tc39.github.io/ecma262/#sec-module-namespace-objects)**
+    - Represents a read-only static set of bindings to a module's exports.
+
+### 2.2. Operations
+
+* **[ParseModule](https://tc39.github.io/ecma262/#sec-parsemodule)**
+    - Creates a [SourceTextModuleRecord]
+    (https://tc39.github.io/ecma262/#sec-source-text-module-records) from
+    source code.
+
+* **[HostResolveImportedModule]
+(https://tc39.github.io/ecma262/#sec-hostresolveimportedmodule)**
+    - A hook for when an import is exactly performed. This returns a
+      `ModuleRecord`. Used as a means to grab modules from Node's loader/cache.
+
+* **[ModuleNamespaceCreate]
+(https://tc39.github.io/ecma262/#sec-modulenamespacecreate)**
+    - Provides a means of creating a list of exports manually, used so that
+      NCJS `module.exports` can create `ModuleRecord`s that are
+      prepopulated.
+
+## 3. Semantics
+
+### 3.1. Determining if source is an ES Module
+
+A new file type will be used for ES Modules, `.mjs`. Files with this extension
+will be treated using the loading semantics in this document, but retain a way
+to load existing extensions of `.json`, `.node`, and `.js`. The file type for
+JavaScript with IANA as an official file type needs to have this extensions
+added. This comes from browsers being unable to implement a new MIME for the
+differing parse goals. In order to make this change, contact the IESG.
+
+The `.mjs` file extension will be searched for prior to any `.js` file in any
+algorithm that uses searching (`require` or `import`); e.g. once the Node
+resolution algorithm reaches file expansion, `path + '.mjs'` would be searched
+prior to `path + '.js'` when performing `require(path)`.
+
+# 3.1.1. Inter package loading using file extension breakage.
+
+There is knowledge of breakage for code that upgrades inner package
+dependencies such as `require('foo/bar.js')`. As `bar.js` may move to
+`bar.mjs`. Since `bar.js` is not the listed entry point this was considered
```

Comment:
Similarly, a common community "best practice" (although not a universal one) is to always omit extensions when requiring - to allow for "bar.js" to become "bar/index.js" transparently, for example - and require('foo/bar') would not break under this proposal.